# Работы с датасетом (начало)

In [1]:
import pandas as pd

Работаем с таблицей о велосипедах

In [2]:
bikes = pd.read_csv('data/BikeData.csv')
print(bikes.shape)
bikes.head()

(8760, 12)


Date  Hour  Temperature  Humidity  Wind speed  Rainfall  Snowfall  \
0  01/12/2017     0         -5.2        37         2.2       0.0       0.0   
1  01/12/2017     1         -5.5        38         0.8       0.0       0.0   
2  01/12/2017     2         -6.0        39         1.0       0.0       0.0   
3  01/12/2017     3         -6.2        40         0.9       0.0       0.0   
4  01/12/2017     4         -6.0        36         2.3       0.0       0.0   

  Seasons     Holiday Functioning Day  Partner 1  Partner 2  
0  Winter  No Holiday             Yes        207         50  
1  Winter  No Holiday             Yes        139         80  
2  Winter  No Holiday             Yes         22        140  
3  Winter  No Holiday             Yes         43        105  
4  Winter  No Holiday             Yes         59         38

In [3]:
bikes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8760 entries, 0 to 8759
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Date             8760 non-null   object 
 1   Hour             8760 non-null   int64  
 2   Temperature      8581 non-null   float64
 3   Humidity         8760 non-null   int64  
 4   Wind speed       8760 non-null   float64
 5   Rainfall         8760 non-null   float64
 6   Snowfall         8760 non-null   float64
 7   Seasons          8760 non-null   object 
 8   Holiday          8760 non-null   object 
 9   Functioning Day  8760 non-null   object 
 10  Partner 1        8760 non-null   int64  
 11  Partner 2        8760 non-null   int64  
dtypes: float64(4), int64(4), object(4)
memory usage: 821.4+ KB


Мы видим, что ненулевых значений в стобце `'Temperature'` не `8760`, а `8581`. Запомним это.

## Фильтрация и сортировка

In [ ]:
bikes['Hour']

In [5]:
bikes['Hour'].unique()  # посмотреть уникальные значения

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

In [6]:
bikes['Functioning Day'].unique()

array(['Yes', 'No'], dtype=object)

In [7]:
bikes[['Partner 1', 'Partner 2']]

Partner 1  Partner 2
0           207         50
1           139         80
2            22        140
3            43        105
4            59         38
...         ...        ...
8755        467        177
8756        296         63
8757        571        665
8758        171        457
8759        442        401

[8760 rows x 2 columns]

In [8]:
bikes.iloc[124]  # вывести 124ую по счёту строку

Date               06/12/2017
Hour                        4
Temperature              -3.4
Humidity                   75
Wind speed                0.0
Rainfall                  0.0
Snowfall                  0.4
Seasons                Winter
Holiday            No Holiday
Functioning Day           Yes
Partner 1                  36
Partner 2                   4
Name: 124, dtype: object

In [9]:
bikes.iloc[124]['Holiday']  # это был не выходной

'No Holiday'

In [10]:
# Сколько выходных часов?
bikes[bikes['Holiday'] == 'Holiday'].shape

(432, 12)

In [12]:
# Сколько выходных часов с температурой > 15?
bikes[(bikes['Holiday'] == 'Holiday') & 
      (bikes['Temperature'] > 15)].shape

(179, 12)

In [15]:
# Сколько выходных часов или/и тех, когда температура > 15?
bikes[(bikes['Holiday'] == 'Holiday') | 
      (bikes['Temperature'] > 15)].shape

(4293, 12)

In [17]:
# Сколько остальных часов? (в которых невыходные и при этом неправда, что температура > 15)
bikes[~((bikes['Holiday'] == 'Holiday') | 
      (bikes['Temperature'] > 15))].shape

(4467, 12)

In [20]:
# Проверка, что посчитали все часы из таблицы. Всего строк в таблице 8760 
4293 + 4467

8760

Почему тут выдаётся другое количество часов?

In [18]:
bikes[(bikes['Holiday'] == 'No Holiday') &  # подумать
      (bikes['Temperature'] <= 15)].shape

(4299, 12)

Потому что сюда не входят часы, когда не выходные и при этом `bikes['Temperature']` имеет нулевое значение `NaN`.

Мы узнали про часы с температурой >15, про часы с температурой <=15, а часы с `NaN` не входят не в первую группу, ни во вторую.

In [22]:
# Сколько всего нулевых значений?
bikes[bikes['Temperature'].isna()].shape

(179, 12)

In [24]:
# Сколько невыходных часов с температурой NaN?
bikes[(bikes['Holiday'] == 'No Holiday') & (bikes['Temperature'].isna())].shape

(168, 12)

In [25]:
# Проверка, что теперь мы учитываем все часы из таблицы. Всего строк в таблице 8760 
4293 + 4299 + 168

8760

**Вывод:**
1. Для отрицания условия используйте `~`
2. Помните, что нулевые значения могут не удовлетворять ни одному из написанных вами условий

До этого мы проверяли соответствие ячейки определённому значению.

Как проверить, что значение в ячейке есть в списке подходящих?

In [28]:
# Вот мы хотим данные только за осень и зиму
bikes[(bikes['Seasons'] == 'Autumn') | (bikes['Seasons'] == 'Winter')].shape

(4344, 12)

In [27]:
# Можно записать проще:
bikes[bikes['Seasons'].isin(['Winter', 'Autumn'])].shape

(4344, 12)

Можно использовать не список, а `range()`

In [29]:
# Таблица только про ночные часы (с полуночи до пяти утра)
bikes[bikes['Hour'].isin(range(0,5))]

Date  Hour  Temperature  Humidity  Wind speed  Rainfall  Snowfall  \
0     01/12/2017     0         -5.2        37         2.2       0.0       0.0   
1     01/12/2017     1         -5.5        38         0.8       0.0       0.0   
2     01/12/2017     2         -6.0        39         1.0       0.0       0.0   
3     01/12/2017     3         -6.2        40         0.9       0.0       0.0   
4     01/12/2017     4         -6.0        36         2.3       0.0       0.0   
...          ...   ...          ...       ...         ...       ...       ...   
8736  30/11/2018     0          1.8        62         1.6       0.0       0.0   
8737  30/11/2018     1          1.3        63         2.0       0.0       0.0   
8738  30/11/2018     2          0.6        66         0.5       0.0       0.0   
8739  30/11/2018     3         -0.1        69         0.9       0.0       0.0   
8740  30/11/2018     4         -0.5        71         0.4       0.0       0.0   

     Seasons     Holiday Functioning Day  Partner 1  Partner 2  
0     Winter  No Holiday             Yes        207         50  
1     Winter  No Holiday             Yes        139         80  
2     Winter  No Holiday             Yes         22        140  
3     Winter  No Holiday             Yes         43        105  
4     Winter  No Holiday             Yes         59         38  
...      ...         ...             ...        ...        ...  
8736  Autumn  No Holiday             Yes        224        109  
8737  Autumn  No Holiday             Yes        110        392  
8738  Autumn  No Holiday             Yes        274         52  
8739  Autumn  No Holiday             Yes         32         84  
8740  Autumn  No Holiday             Yes         48         28  

[1825 rows x 12 columns]

Будьте осторожны, `range()` работает с целыми числами.

То есть, если вы смотрите температуру в `range(10, 15)`, вы находите часы с температурой `10`, `11`, `12`, `13`, `14`, но не `10.5`, например.

In [32]:
bikes[bikes['Temperature'].isin(range(10, 15))].shape  # это для целых чисел

(100, 12)

### Сортировка

In [33]:
# Отсортируем температуру
bikes['Temperature'].sort_values()

1352   -17.8
1353   -17.5
1351   -17.5
1350   -17.4
1349   -16.9
        ... 
8669     NaN
8679     NaN
8710     NaN
8717     NaN
8742     NaN
Name: Temperature, Length: 8760, dtype: float64

In [34]:
# Отсортируем температуру по убывающей
bikes['Temperature'].sort_values(ascending=False)

5848    39.4
5847    39.3
5849    39.0
5846    38.7
5825    38.0
        ... 
8669     NaN
8679     NaN
8710     NaN
8717     NaN
8742     NaN
Name: Temperature, Length: 8760, dtype: float64

Заметим, что `NaN` всегда в конце.

Как отсортировать всю таблицу по столбцу с температурой?

In [35]:
bikes.sort_values(by=['Temperature'])

Date  Hour  Temperature  Humidity  Wind speed  Rainfall  Snowfall  \
1352  26/01/2018     8        -17.8        40         2.5       0.0       0.5   
1353  26/01/2018     9        -17.5        37         3.8       0.0       0.5   
1351  26/01/2018     7        -17.5        38         3.4       0.0       0.5   
1350  26/01/2018     6        -17.4        40         1.6       0.0       0.5   
1349  26/01/2018     5        -16.9        38         2.3       0.0       0.5   
...          ...   ...          ...       ...         ...       ...       ...   
8669  27/11/2018     5          NaN        84         0.6       0.0       0.0   
8679  27/11/2018    15          NaN        70         2.1       0.0       0.0   
8710  28/11/2018    22          NaN        34         1.3       0.0       0.0   
8717  29/11/2018     5          NaN        34         0.9       0.0       0.0   
8742  30/11/2018     6          NaN        68         0.8       0.0       0.0   

     Seasons     Holiday Functioning Day  Partner 1  Partner 2  
1352  Winter  No Holiday             Yes        235        288  
1353  Winter  No Holiday             Yes         76         26  
1351  Winter  No Holiday             Yes         93         31  
1350  Winter  No Holiday             Yes         33         18  
1349  Winter  No Holiday             Yes         35         21  
...      ...         ...             ...        ...        ...  
8669  Autumn  No Holiday             Yes        118         58  
8679  Autumn  No Holiday             Yes        391        279  
8710  Autumn  No Holiday             Yes        476        558  
8717  Autumn  No Holiday             Yes         17         13  
8742  Autumn  No Holiday             Yes         11        155  

[8760 rows x 12 columns]

In [36]:
# А теперь по нескольким столбцам
bikes.sort_values(by=['Temperature', 'Humidity'])

Date  Hour  Temperature  Humidity  Wind speed  Rainfall  Snowfall  \
1352  26/01/2018     8        -17.8        40         2.5       0.0       0.5   
1353  26/01/2018     9        -17.5        37         3.8       0.0       0.5   
1351  26/01/2018     7        -17.5        38         3.4       0.0       0.5   
1350  26/01/2018     6        -17.4        40         1.6       0.0       0.5   
1349  26/01/2018     5        -16.9        38         2.3       0.0       0.5   
...          ...   ...          ...       ...         ...       ...       ...   
5114  02/07/2018     2          NaN        97         0.8       2.0       0.0   
5336  11/07/2018     8          NaN        97         0.8       0.0       0.0   
6526  29/08/2018    22          NaN        97         2.1       4.5       0.0   
2506  15/03/2018    10          NaN        98         1.2       0.0       0.0   
6463  27/08/2018     7          NaN        98         0.5       2.0       0.0   

     Seasons     Holiday Functioning Day  Partner 1  Partner 2  
1352  Winter  No Holiday             Yes        235        288  
1353  Winter  No Holiday             Yes         76         26  
1351  Winter  No Holiday             Yes         93         31  
1350  Winter  No Holiday             Yes         33         18  
1349  Winter  No Holiday             Yes         35         21  
...      ...         ...             ...        ...        ...  
5114  Summer  No Holiday             Yes          9         14  
5336  Summer  No Holiday             Yes        537        384  
6526  Summer  No Holiday             Yes          0         17  
2506  Spring  No Holiday             Yes         27         19  
6463  Summer  No Holiday             Yes         73        103  

[8760 rows x 12 columns]

In [37]:
# Температуру по возрастанию, а влажность по убыванию
bikes.sort_values(by=['Temperature', 'Humidity'], ascending=[True, False])

Date  Hour  Temperature  Humidity  Wind speed  Rainfall  Snowfall  \
1352  26/01/2018     8        -17.8        40         2.5       0.0       0.5   
1351  26/01/2018     7        -17.5        38         3.4       0.0       0.5   
1353  26/01/2018     9        -17.5        37         3.8       0.0       0.5   
1350  26/01/2018     6        -17.4        40         1.6       0.0       0.5   
1349  26/01/2018     5        -16.9        38         2.3       0.0       0.5   
...          ...   ...          ...       ...         ...       ...       ...   
1238  21/01/2018    14          NaN        18         1.2       0.0       0.0   
1910  18/02/2018    14          NaN        17         1.8       0.0       0.0   
1886  17/02/2018    14          NaN        15         4.0       0.0       0.0   
3400  21/04/2018    16          NaN        13         3.3       0.0       0.0   
4276  28/05/2018     4          NaN         0         0.4       0.0       0.0   

     Seasons     Holiday Functioning Day  Partner 1  Partner 2  
1352  Winter  No Holiday             Yes        235        288  
1351  Winter  No Holiday             Yes         93         31  
1353  Winter  No Holiday             Yes         76         26  
1350  Winter  No Holiday             Yes         33         18  
1349  Winter  No Holiday             Yes         35         21  
...      ...         ...             ...        ...        ...  
1238  Winter  No Holiday             Yes        119        100  
1910  Winter  No Holiday             Yes          7        162  
1886  Winter     Holiday             Yes          0         58  
3400  Spring  No Holiday             Yes       1649        560  
4276  Spring  No Holiday             Yes         51         58  

[8760 rows x 12 columns]

Что влияет на количество взятых велосипедов? Наличие осадков? Температура воздуха? Выходной ли день? Это всё можно узнать.

Для удобства подсчётов создадим новый столбик с общим количеством арендованных велосипедов.

In [39]:
bikes['Rental Count'] = bikes['Partner 1'] + bikes['Partner 2']

In [40]:
bikes.head()

Date  Hour  Temperature  Humidity  Wind speed  Rainfall  Snowfall  \
0  01/12/2017     0         -5.2        37         2.2       0.0       0.0   
1  01/12/2017     1         -5.5        38         0.8       0.0       0.0   
2  01/12/2017     2         -6.0        39         1.0       0.0       0.0   
3  01/12/2017     3         -6.2        40         0.9       0.0       0.0   
4  01/12/2017     4         -6.0        36         2.3       0.0       0.0   

  Seasons     Holiday Functioning Day  Partner 1  Partner 2  Rental Count  
0  Winter  No Holiday             Yes        207         50           257  
1  Winter  No Holiday             Yes        139         80           219  
2  Winter  No Holiday             Yes         22        140           162  
3  Winter  No Holiday             Yes         43        105           148  
4  Winter  No Holiday             Yes         59         38            97

### Тренировка

Когда больше берут велосипедов? Осенью и зимой или весной и летом?

In [ ]:
bikes[bikes['Seasons'].isin(['Winter', 'Autumn'])]['Rental Count'].sum()

In [ ]:
bikes[bikes['Seasons'].isin(['Spring', 'Summer'])]['Rental Count'].sum()

Что это значит?

За все летние и весенние дни велосипедов взяли больше, чем за все осенние и зимние.

In [ ]:
bikes[bikes['Seasons'].isin(['Winter', 'Autumn'])]['Rental Count'].mean()

In [ ]:
bikes[bikes['Seasons'].isin(['Spring', 'Summer'])]['Rental Count'].mean()

Что это значит?

Осенью и зимой в среднем берут 518.7 велосипедов в час, а весной и летом 871.6 велосипедов в час. То есть весной и летом чаще.

**Почему стоит в такой задаче использовать `.mean()`, а не `.sum()`?**

Представьте, что у нас иформация про 100 зимних дней и про 1 летний. Летом гораздо чаще берут велосипеды. Но суммарно за 100 дней зимы велосипедов могут взять гораздо больше, чем за один летний.

```
Например: 50 человек в городе. Зимой берут велосипеды два человека в день. Летом берут все 50 человек.
- за все летние дни (у нас данные про один): 50 велосипедов (50 велосипедов * 1 день)
- за все зимние дни (у нас данные про сто): 200 велосипедов (2 велосипеда * 100 дней)

Средние арифметические показывают, что летом берут чаще, чем зимой: 50>2. А общие суммы не показывают.
```

### Тренировка

Что больше влияет на то, берут ли велосипеды: температура или выходной-не_выходной?

**Про температуру:**

In [ ]:
bikes[bikes['Temperature'] > 15]['Rental Count'].mean()

In [ ]:
bikes[bikes['Temperature'] <= 15]['Rental Count'].mean()

Получается, что чаще берут велосипеды в теплые часы.

**Про выходные:**

In [ ]:
bikes[bikes['Holiday'] == 'No Holiday']['Rental Count'].mean()

In [ ]:
bikes[bikes['Holiday'] == 'Holiday']['Rental Count'].mean()

Получается, что чаще берут велосипеды в будни, а не в выходные.

**Итог:** поскольку разница между 1021 и 410 больше, чем между 708 и 470, получается, что при выборе брать ли велосипед температура важнее, чем тип дня (выходной или нет).